images/plugin-compiler/testplugin/main.go

```diff
@@ -0,0 +1,239 @@
+		r.Header.Add("current_time", now)
+		return
+	}
+
+	// send HTTP response from Golang plugin
+	logger.WithField("api-name", apidef.Name).Info("Golang plugin - Setting the response header and body. Request will stop in this plugin and OK response will be returned")
+
+	// prepare data to send
+	replyData := map[string]string{
+		"current_time": now,
+	}
+
+	rw.Header().Set("Content-Type", "application/json")
+	rw.WriteHeader(http.StatusOK)
+	writeBody(rw, replyData)
+}
+
+func writeBody(rw http.ResponseWriter, replyJson map[string]string) error{
+
+	var logger = log.Get()
+	jsonData, err := json.Marshal(replyJson)
+	if err != nil {
+		logger.WithField("replyJson", replyJson).Error("Golang auth plugin: Failed to marshal")
+		rw.WriteHeader(http.StatusInternalServerError)
+		return err
+	}
+
+	rw.Write(jsonData)
+	return nil
+}
+
+
+// ------------------------------------------------------------------
+// Custom auth plugin code:
+
+func getSessionByKey(key string) *user.SessionState {
+	// here goes our logic to check if passed API key is valid and appropriate key session can be retrieved
+
+	// perform auth (only one token "abc" is allowed)
+	// Here you add code to query your database
+	if key != "abc" {
+		return nil
+	}
+
+	// return session
+	return &user.SessionState{
+		OrgID: "default",
+		Alias: "abc-session",
+	}
+}
+
+func MyPluginCustomAuthCheck(rw http.ResponseWriter, r *http.Request) {
+
+	var logger = log.Get()
+	apidef := ctx.GetDefinition(r)
+	logger.WithField("api-name", apidef.Name).Info("Golang auth plugin - MyPluginCustomAuthCheck")
+
+	// try to get session by API key
+	key := r.Header.Get(headers.Authorization)
+	session := getSessionByKey(key)
+	if session == nil {
+		// auth failed, reply with 403
+		logger.WithField("api-name", apidef.Name).Info("Golang auth plugin - MyPluginCustomAuthCheck - failed")
+		rw.WriteHeader(http.StatusForbidden)
+
+		// prepare data to send
+		replyData := map[string]string{
+			"reason": "Access forbidden",
+		}
+
+		writeBody(rw, replyData)
+		//jsonData, err := json.Marshal(replyData)
+		//if err != nil {
+		//	rw.WriteHeader(http.StatusInternalServerError)
+		//
+		//	rw.Write(jsonData)
+		//	return
+		//}
+
+		return
+	}
+
+	logger.WithField("api-name", apidef.Name).Info("Golang auth plugin - MyPluginCustomAuthCheck - succeeded")
+
+
+	// auth was successful, add session and key to request's context so other middle-wares can use it
+	ctx.SetSession(r, session, key, true)
+}
+
+type OauthClientRequest struct {
+	Method string `json:"method" msg:"method"`
+	Body   string `json:"body" msg:"body"`
+	Headers   string `json:"body" msg:"body"`
+	Domain   string `json:"domain" msg:"domain"`
+	Resource   string `json:"resource" msg:"resource"`
+
+}
+func main() {}
+
+
+/*
+func CustomAuthPluginOAuthOkta(rw http.ResponseWriter, r *http.Request)  () {
```

Comment:
Feels like we can remove this dead code?

Reply:
it's not even okta, it looks like auth0